iceoryx2/tests/waitset_tests.rs

```diff
@@ -0,0 +1,466 @@
+
+        let (listener_1, notifier_1) = create_event::<S>(&node);
+        let (listener_2, _notifier_2) = create_event::<S>(&node);
+        let (receiver_1, sender_1) = create_socket();
+        let (receiver_2, _sender_2) = create_socket();
+
+        let listener_1_guard = sut.attach_notification(&listener_1).unwrap();
+        let listener_2_guard = sut.attach_notification(&listener_2).unwrap();
+        let receiver_1_guard = sut.attach_notification(&receiver_1).unwrap();
+        let receiver_2_guard = sut.attach_notification(&receiver_2).unwrap();
+
+        notifier_1.notify().unwrap();
+        sender_1.try_send(b"bla").unwrap();
+
+        let mut listener_1_triggered = false;
+        let mut listener_2_triggered = false;
+        let mut receiver_1_triggered = false;
+        let mut receiver_2_triggered = false;
+
+        sut.run(|attachment_id| {
+            if attachment_id.event_from(&listener_1_guard) {
+                listener_1_triggered = true;
+            } else if attachment_id.event_from(&listener_2_guard) {
+                listener_2_triggered = true;
+            } else if attachment_id.event_from(&receiver_1_guard) {
+                receiver_1_triggered = true;
+            } else if attachment_id.event_from(&receiver_2_guard) {
+                receiver_2_triggered = true;
+            } else {
+                test_fail!("only attachments shall trigger");
+            }
+        })
+        .unwrap();
+
+        assert_that!(listener_1_triggered, eq true);
+        assert_that!(receiver_1_triggered, eq true);
+    }
+
+    #[test]
+    fn run_with_tick_blocks_for_at_least_timeout<S: Service>()
+    where
+        <S::Event as Event>::Listener: SynchronousMultiplexing,
+    {
+        let _watchdog = Watchdog::new();
+        let node = NodeBuilder::new().create::<S>().unwrap();
+        let sut = WaitSetBuilder::new().create::<S>().unwrap();
+
+        let (listener, _) = create_event::<S>(&node);
+        let _guard = sut.attach_notification(&listener);
+        let tick_guard = sut.attach_tick(TIMEOUT).unwrap();
+
+        let mut callback_called = false;
+        let start = Instant::now();
+        sut.run(|id| {
+            callback_called = true;
+            assert_that!(id.event_from(&tick_guard), eq true);
+            assert_that!(id.deadline_from(&tick_guard), eq false);
+        })
+        .unwrap();
+
+        assert_that!(callback_called, eq true);
+        assert_that!(start.elapsed(), time_at_least TIMEOUT);
+    }
+
+    #[test]
+    fn run_with_deadline_blocks_for_at_least_timeout<S: Service>()
+    where
+        <S::Event as Event>::Listener: SynchronousMultiplexing,
+    {
+        let _watchdog = Watchdog::new();
+        let node = NodeBuilder::new().create::<S>().unwrap();
+        let sut = WaitSetBuilder::new().create::<S>().unwrap();
+
+        let (listener, _) = create_event::<S>(&node);
+        let guard = sut.attach_deadline(&listener, TIMEOUT).unwrap();
```

Comment:
Same idea here, although it becomes a little bit long and I'm not even sure I like it. 
```suggestion
        let guard = sut.attach_deadline_interval(&listener, TIMEOUT).unwrap();
```